### ISSUE TITLE
DS: Enable Python Interactive for use as a debugging console

### ISSUE THREAD

**Issue by jbeyer16** [microsoft/vscode-jupyter#1278]
I think this is related to / mentioned deep in the comments [here](https://github.com/microsoft/vscode-python/issues/5113#issuecomment-481300489), but I'm not convinced they're quite duplicates.

I think it'd be tremendously useful to be able to also select Python Interactive as your debug terminal, essentially launching a Python Interactive window at a breakpoint with the state of your program loaded into it, giving you access to your variables. That way plotting functions and visualizations could be made use of pretty easily as a way of debugging. This varies slightly from the current "Debug Current File in Python Interactive window", which would run your entire script in Python Interactive, in that you are still able to configure the debugger launch.json. For my specific use case it would allow passing in command line args to the script, while still being able to debug in Python Interactive.

At a high level, I can see two ways of it happening:
1) Use only Python Interactive and allow it to continue execution of the code as the full debug tool.
2) Use the main debug terminal for the debugging flow, but at a breakpoint launch the Python Interactive window that opens preloaded with the current program state. At that point, the Python Interactive window is essentially detached from debugger, but at least has the data from the program at that time, allowing for more native exploring of the variables. Once the user continues debugging, the current Python Interactive window is either updated or a new one is launched at the next breakpoint. 

My current work around is just to write a data frame to a .csv in the current debugger, then load it into Python Interactive - which works well enough, but I think it'd be really helpful if they were somehow integrated!

Thank you for making this product so awesome to use!

**Comment by Steph71** [comment]
Hi, 

I am using VScode Python for remote debugging, because Pycharm blocks this feature in its community edition.

I would love to make VScode my main IDE for Python, because I like the idea to have one Editor for all languages.

But the missing interactive debug console is a showstopper for me. 

Hitting a breakpoint and dropping into an IPython Console is an essential feature for a Python IDE:

https://www.jetbrains.com/help/pycharm/using-debug-console.html

Actually also the good old pydev can do it:

https://www.pydev.org/manual_adv_debug_console.html

Best Regards

**Comment by rchiodo** [comment]
@Steph71 not sure what you're missing. The IPython console is available when using the interactive window for debugging. Click the 'Debug Cell' code lens over top of your code. All output will be sent to the interactive window.

![image](https://user-images.githubusercontent.com/19672699/100906189-c1170600-347d-11eb-9fcc-53b91be54e60.png)


Perhaps you're talking about being able to eval in the ipython console. That's not currently supported (as the kernel is busy with the cell that's running). You can eval in the debug console at the same time though. Output is then duplicated in the interactive window and the debug console:

![image](https://user-images.githubusercontent.com/19672699/100906964-9e392180-347e-11eb-8f0e-961859940c28.png)

**Comment by rchiodo** [comment]
Or perhaps you're asking for a switch to 'interactive' after you've already started debugging. This won't likely be supported any time soon (I mentioned this in one of the comments above). We'd have to transfer the state of the python app you launched into a jupyter kernel.

**Comment by jbeyer16** [comment]
Maybe it's worth restating what I thought when I created this issue and what I've learned thanks to some comments from others.

The initial hope was to have an additional option in the `launch.json` file under `console` that allowed you to pick "Python Interactive". The idea was then, when a debug breakpoint was hit, instead of working in the debug console, a Python Interactive window would open with your variables in the workspace, and you could play around in Python Interactive as you desired. I have no idea if this is feasible or not, but sounds not particularly likely.

In lieu of this feature, there was a great suggestion above to run:

`from IPython import embed; embed()`

in the Debug Console as soon as a breakpoint hit. You could then click "Terminal" and have an embedded IPython console running with all your workspace variables available. There are some differences between the IPython console and Interactive Python, but functionally they work pretty similarly. This method accomplishes pretty much everything I was hoping for, just with a bit of manual interaction and without the friendliness of the Python Interactive interface. 

You can then, for example, run (in your embedded IPython console):

`import matplotlib.pyplot as plt`
`plt.ion()` 

Then you can then create and explore plots with your data from the state of your script at the debug breakpoint. The downside here is that the embedded IPython console essentially blocks the debugger, so you can't interact with any of the Variables on the VSCode side until you terminate the IPython session. Even still, I think this is a very helpful alternative.

So one suggestion may be just to automate that `from IPython import embed; embed()` call. So maybe in the `launch.json` file, you can select "IPython Terminal" as your console, and basically behind the scenes that line of code runs for you and pops up the IPython console. Not sure. But like I said, running `from IPython import embed; embed()` accomplishes most all of what I'm looking for, and I encourage others to try that out and see what they think!

**Comment by rchiodo** [comment]
> The idea was then, when a debug breakpoint was hit, instead of working in the debug console, a Python Interactive window would open with your variables in the workspace, and you could play around in Python Interactive as you desired. I have no idea if this is feasible or not, but sounds not particularly likely.

Actually this idea is pretty feasible. Switching to this console after starting isn't, but starting the debuggee with the interactive window should be doable. It would essentially run your entire process as a cell.

In lieu of that, your other workaround sounds good too.

**Comment by Steph71** [comment]
Hi,

thanks for all the answers.

I tried to run my code in the interactive window ... but I do not see, how I would set a breakpint there and step in and through functions.

the suggested way

`from IPython import embed; embed()`

does basically what I am looking for, but it has some downsides:

1.) I have to type or paste the command whenever I need the interactive console, then I have to switch windows
2.) The main debug process is stalled, so I if I want to step one line in my original code, I need to exit the Ipython console, do the step, then start the console again ...
3.) The pycharm debug console is giving me nice features like 
 - its just allways there, when I hit a breakpoint
 - syntax highlighting
 - it remembers my commonly used functions, and properties for a specific datatype and puts them on top of the tab completion list

to me this sounds more like a workaround than a feature ... but this is probably just the difference bewteen a generic editor, enabled with some plugin to support a language and an IDE, specifically build to develop in a language ...

Best Regards
Stephan

**Comment by hoechenberger** [comment]
@rchiodo 
> but starting the debuggee with the interactive window should be doable. It would essentially run your entire process as a cell.

This would be exactly what I'm looking for :)

**Comment by rchiodo** [comment]
@Steph71 
> I tried to run my code in the interactive window ... but I do not see, how I would set a breakpint there and step in and through functions.

Setting a breakpoint is the same as setting breakpoints without the interactive window. You click in the gutter or hit F9.

![image](https://user-images.githubusercontent.com/19672699/101064674-1f5ceb00-3549-11eb-8cb5-b3bcbc6f7f26.png)

As you step of if you hit F5, the breakpoints should hit. This should also work with breakpoints in other files. It's running the same debugger as non interactive.

> but this is probably just the difference bewteen a generic editor, enabled with some plugin to support a language and an IDE, specifically build to develop in a language ...

I agree it's not ideal. We should implement what @jbeyer16 asked for, set the console for any debug session to 'interactive' and then it just runs the entire file you have open in the interactive window. This would also require changes to the interactive window so you could input during debugging (the mismatch you may be referring to between a full IDE and an editor). 

I'd argue that VS code is an IDE, it's just the scenario you're asking for wasn't implemented yet.

**Comment by Steph71** [comment]
@rchiodo

ah, now I got it debug ... but actually its the same issue here ... while debugging the interactive window is not interactive but frozen.

If the option @jbeyer16 is proposing is feasable, that would be a big step forward ... actually the  Interavtive window is even somehwhat better than the ipython console as plots can be directly plotted to the window instead having them floating around ...

So I will watch this issue and hope for that getting resolved the one or the other way ... probably its worth looking how the pydev people are doing that, as jetbrains are just using their code ...

**Comment by OSHI7** [comment]
I'd second @jbeyer16 's suggestion.

**Comment by oponcea-dn** [comment]
This would be a great feature! I just started using vscode debugging and got used to PyCharm way of doing things. So, any news on it?

**Comment by rchiodo** [comment]
It's on our backlog but no news on when we'll get to it.

**Comment by sunshe35** [comment]
i love it.
i love debug python file using ipython console like pycharm.
so i do not like the debuging way for python now when i use vs-code.

**Comment by ChelsyMena** [comment]
Hey, correct me if I'm wrong but is this not the "right click + open a console for notebook" function in JupyterLab? That's what I miss the most when I'm in VSC (and Shift + Tab docstrings, but that's another beast)

**Comment by aksg87** [comment]
@JanHomann @jbeyer16

When following the steps of launching: `from IPython import embed; embed()` 

I am not getting a second external window pop up after trying to display my plot. Anything I am missing here? 

My steps:
1. `from IPython import embed; embed()` 
2.  `import matplotlib.pyplot as plt`
3. `plt.imshow(...)` + `plt.show()`

![image](https://user-images.githubusercontent.com/19271735/128558285-39902cb8-7413-4344-8074-74d5f96197d7.png)

**Comment by aksg87** [comment]
> @Steph71 not sure what you're missing. The IPython console is available when using the interactive window for debugging. Click the 'Debug Cell' code lens over top of your code. All output will be sent to the interactive window.
> 
> ![image](https://user-images.githubusercontent.com/19672699/100906189-c1170600-347d-11eb-9fcc-53b91be54e60.png)
> 
> Perhaps you're talking about being able to eval in the ipython console. That's not currently supported (as the kernel is busy with the cell that's running). You can eval in the debug console at the same time though. Output is then duplicated in the interactive window and the debug console:
> 
> ![image](https://user-images.githubusercontent.com/19672699/100906964-9e392180-347e-11eb-8f0e-961859940c28.png)

When trying to debug an individual cell, I still can't interrogate data and plot new aspects while debugging the cell. The debug cell does allow me to step through the existing code and see variables states, but I still don't see how I can plot new things dynamically?

**Comment by jbeyer16** [comment]
@aksg87 I think you might have to create the figure explicitly before plotting.

```python
from IPython import embed; embed()
import matplotlib.pyplot as plt
plt.ion()
plt.figure()
plt.plot([1,2], [3,3])
```

seems to work for me.

**Comment by vintprox** [comment]
In my use case, it's one keybinding away on `Jupyter: Debug Current File in Interactive Window` command, but I wouldn't like to waste students' time to set these up. Alas, no default is provided.
`launch.json` only allows old-school terminal with big payload (that one is distracting, for when you press <kbd>F5</kbd>).

**Comment by TheSamFed** [comment]
Just created a bounty on Bountysource for this issue, would really love to see this as a feature:
https://app.bountysource.com/issues/98730134-enable-python-interactive-for-use-as-a-debugging-console

**Comment by landisrm** [comment]
It seems like there are a lot of different possibilities discussed here, but one of them seems to be: use IPython in the debug console.  I would sure love to see this feature as well.  I don't want to debug in a Jupyter type interface.  I just want to set a breakpoint and when I hit it, I can use IPython to inspect things.  Maybe this is already possible but if so the setup is not well documented.  

As others have said, this is a major limitation of using VS Code for Python, esp. data science type tasks

**Comment by ddfabbro** [comment]
> In my use case, it's one keybinding away on `Jupyter: Debug Current File in Interactive Window` command, but I wouldn't like to waste students' time to set these up. Alas, no default is provided. `launch.json` only allows old-school terminal with big payload (that one is distracting, for when you press F5).

How can I keep using the interactive terminal while debugging the current file in the Interactive Window?

![image](https://user-images.githubusercontent.com/17279638/175860999-3b83ee23-a722-4196-add8-675092f55b10.png)

I have a situation where during a debugging session I would like to interact with the images using matplotlib.

Problems:
- There is still a long process after that, so I can not wait for the current cell to finish.
- Also my code base is fairly large, so I can not start splitting it into Jupyter cells.

The way I am dealing with this is logging the images to file (.jpg, .png, etc) in an oldschool OpenCV C++ style. That is not interactive at all, and I always need to convert the images to unsigned chars before saving it to .jpg. Matploblib lets me visualize arrays with any datatype as image using `plt.imshow`

**Comment by rchiodo** [comment]
You simply type your expressions into the debug console instead of the edit box for the interactive window.

Internally the interactive window is listening to the output of a cell running. So we do something like this:

```
    const request = kernel.requestExecute(`Your code that submitted for the entire cell');
    request.onOutput((o) => this.sendOutputToView(o))
```

So it won't accept another request (which is what the input box sends).

However the debug console window does something different. There's a separate thread in the kernel that the debugger inserts and it evaluates whatever you type on that thread. So it's not blocked.

However the output from that eval ends up getting sent back to our code as an output.

Here's an example of me printing something from the debug console as I'm debugging a cell:

![debuggerconsole](https://user-images.githubusercontent.com/19672699/175989589-1ac5050f-58db-4d85-a574-e98c1bd6f76f.gif)

**Comment by kleffy** [comment]
Sorry! is the feature suggested by @jbeyer16 available now?

**Comment by kleffy** [comment]
Actually, I just found out that the feature exist in vs code. see solution from this [stackoverflow](https://stackoverflow.com/questions/70878545/vs-code-how-to-launch-an-interactive-python-terminal-while-debugging) thread.

**Comment by hoechenberger** [comment]
> Actually, I just found out that the feature exist in vs code. see solution from this [stackoverflow](https://stackoverflow.com/questions/70878545/vs-code-how-to-launch-an-interactive-python-terminal-while-debugging) thread.

This is not what this question / feature request is about. @jbeyer16 was not asking for **any** interactive debugger, but for the debugger to run in what the Python extension calls the Python Interactive console.

**Comment by roblourens** [comment]
I'm wondering, would that still work for outputs? Yesterday Don said that running code via the debugger does not run in "inside the kernel", and I don't know exactly what he means by that, but maybe outputs are what would be missing.

**Comment by rchiodo** [comment]
As far as I remember, running evals in the debugger thread still causes the output to go to the currently running cell.

You can test this by just evaling something in the debug console while debugging the interactive window.

**Comment by TBlackmore** [comment]
Another vote for this feature!

**Comment by raphaelchinchilla** [comment]
It has been more than 4 years since this post appeared. Any update on when it is going to be implemented? 

One thing that (hopefully) is easy to implement is to have a better support for the %pdb magic command of IPython

**Comment by v4lue4dded** [comment]
I just published package that is my attempt at solving this problem.

The idea is to make it easy to save all local and global variables at a specific point during the debug process to disc and then load them again into an interactive window (and even close the interactive window and load them again and again if needed).

https://github.com/v4lue4dded/debugsnap/
https://pypi.org/project/debugsnap/

first install the package into the env you want to debug:
```bash
pip install debugsnap
```

then start a debugging session using the interactive window environment where you want to debug:
![image](https://github.com/microsoft/vscode-jupyter/assets/61813337/547dfbc9-1c3c-4772-b9c0-8f2dba4d2181)

(this doesn't have a keyboard shortcut by default, `Ctrl +Shift + Enter` is my recommendation.

Go a breakpoint in the code where you want to save the snapshot and there open the debug consol:

![image](https://github.com/microsoft/vscode-jupyter/assets/61813337/3ff6722a-51b0-4cc3-8f20-36ff0b4cd63d)

1. **Saving the Debugging Snapshot**:
   In your debug session, save the current snapshot.
   ```python
    from debugsnap import save_snapshot
    save_snapshot(local_vars=locals(), global_vars=globals())
   ```

> ** Warning:** By default a new folder named `tmp_save_snapshot` will be created in your `~` directory and the shapshot files (containing the **global variables**, **local variables** and **information** about any **variable that could not be saved** for any reason) will be saved there. To change the storage location you have to provide a `storage_path`.

2. **Loading the Debugging Snapshot**:
   In a new Python session or an interactive environment, load the saved snapshot:
   ```python
    from debugsnap import load_snapshot
    global_vars, local_vars = load_snapshot()
    globals().update(global_vars)
    locals().update(local_vars)
   ```

Now you have all the variables that where present at the breakpoint in an interactive window.

PS: follow the links for info on how to make the saving and loading code into VSCode **snippets**

**Comment by b-a0** [comment]
> We'll consider it in the next triage meeting.

@IanMatthewHuff I understand it's a while ago, but do you remember what the outcome was in the triage meeting?

**Comment by b-a0** [comment]
Sorry to bother you again @IanMatthewHuff, but did you have the chance to check what the outcome was of the last triage meeting? Or are you no longer involved and is it better to contact someone else?

**Comment by amunger** [comment]
Ian is not on the project anymore.

With the number of upvotes, this is something we would love to implement, it's just hard to find the time.

It would be great to see more feedback on @v4lue4dded's solution [posted above](https://github.com/microsoft/vscode-jupyter/issues/1278#issuecomment-1847012815) to know if that is in the right direction.

**Comment by raphaelchinchilla** [comment]
@b-a0 I don't know if this will help:

After some growing frustration with the interactive mode not being similar enough to Spyder IDE, I decided to make VS Code more similar to it by myself.

- Install ipython and some nice extra stuff by doing `pip install ipykernel`
- I start my development by simply typing `ipython` in a terminal.
- I created a shortcut that sends the selected text to the terminal and another one that execute %run name_of_the_file (ask chatgpt help if you need. I say that because I did it)
- I set breakpoints by writing `breakpoint` in the code. When you hit a breakpoint, you should get a ipdb debugger instance. When you are in the debugging, just type `interact` and you have a full ipython console. The cool thing is that all of the shortcuts I defined also just work.
- The only thing I was never able to figure out was a way to have plots being open by vs code. The plots now just open as other windows, which does not work for remote development.

I don't want to criticize the interactive mode, but as @amunger themselves said, they do not have time to develop this debugging feature, which for me is essential. I spent most of my time developing in debugging mode and I want the same dynamic that one gets of of IPython.

Also, in another point instead of improving the interactive mode, the Python team of VS Code (which not the same as the Jupyter team of VS Code, which developed the Interactive mode) has been dedicating resources to create an integrated REPL that (last time I checked ) does not address many of the issues that the community seems to have with the interactive mode, including the one on this thread...

In my opinion, IPython is a phenomenal project, with very good people working on it, and is the best REPL for Python. The vanilla version is not as complete, but is super powerful.

**Comment by b-a0** [comment]
> Ian is not on the project anymore.
Thanks for clarifying and sorry for the bother to Ian.

> It would be great to see more feedback on @v4lue4dded's solution [posted above](https://github.com/microsoft/vscode-jupyter/issues/1278#issuecomment-1847012815) to know if that is in the right direction.

In spirit this is the functionality, but `debugsnap` is a bit clunky to use, especially if I'm working on multi-GB variables that need to be written out and loaded in.

Perhaps it's a slightly silly/rude description, but I think the best way of showing what I am looking for (I hope it matches what this issue is about) is to [download Spyder](https://github.com/spyder-ide/spyder/releases), write a script, set a breakpoint and start debugging.

**Comment by dempc6469** [comment]
> [@raphaelchinchilla](https://github.com/raphaelchinchilla) great to see that you found this creative workaround. Unfortunately I have to share my code with others and it would introduce an unacceptable overhead if I would scatter breakpoint commands everywhere and/or would have to remove those before each commit.

same .. this is such an unfortunate feature miss. Will utilize the workaround for now but it's certainly not ideal.

**Comment by raphaelchinchilla** [comment]
For all that matters, I believe there is a pre-commit hook that will check for that, but I am not sure. If anyone finds please share if possible :).

Also, I don't know if I have a configuration problem in my VS Code, but running the debugger makes my code orders of magnitude slower than when I use a breakpoint.

**Comment by sidhanthapoddar99** [comment]
![img](https://github.com/sidhanthapoddar99/pyrept-notebook-vscode-debug-extension/blob/master/icons/dnb-icon-large.png?raw=true)


I faced this similar problem, wanted to run couple of lines of code in interactive mode for a flask server, so created an extension as a weekend project. 

Is very basic right now but might be helpful

https://marketplace.visualstudio.com/items?itemName=sidh1999.debug-notebook&ssr=false#overview

debug-notebook vscode marketplace

basically, using vscode api to call the debug console, this kind of acts like a wrapper and gives an interactive interface

I have open sourced the project so feel free to contribute and share your thoughts

works for `JavaScript` and `Python`

https://github.com/[sidhanthapoddar99/pyrept-notebook-vscode-debug-extension](https://github.com/sidhanthapoddar99/pyrept-notebook-vscode-debug-extension)

**Comment by RobinL** [comment]
The closest I've got to the desired functionality is as follows, but it's seems very counterintuitive/difficult to discover:

I generally use .py files and execute 'cells' by selecting code and pressing Shift+Enter.

In one of these Python files, I use the Jupyter: Debug current file in interactive window command.

At first, this doesn't seem to do anything special - you still execute debug commands in the debug console, making it appear no different from standard debugging.

But if you wrap any command in display(), the output also appears in the interactive window with full rich formatting.

This gives you the best of both worlds: interactive debugging with the ability to visualize complex data structures and plots in the same workflow.

One reason it's not very intuitive is you can't write statements in the interactive window to access variables in the debug session, you have to write everything in the terminal and wrap things in 'display' to see the interactive output

**Comment by joshdunnlime** [comment]
@RobinL great recommendation. Really useful to know and a bit of a shame that it doesn't work in the interactive window. It's like they failed at the very last step 😅

**Comment by raphaelchinchilla** [comment]
What is even more weird is that if one puts a `breakpoint()` it stops and one can put commands, but they are on a window on the top of the screen.

**Comment by zmbc** [comment]
Great recommendation @RobinL! It seems like there is no deep technical reason why this can't happen, I'm curious if anyone has taken a look into how it would be implemented

**Comment by sidhanthapoddar99** [comment]
> Great recommendation [@RobinL](https://github.com/RobinL)! It seems like there is no deep technical reason why this can't happen, I'm curious if anyone has taken a look into how it would be implemented

https://github.com/microsoft/vscode-jupyter/issues/1278#issuecomment-2822622982

Not sure if this is what you're looking for, but I built a VS Code extension called Debug Notebook.
https://marketplace.visualstudio.com/items/?itemName=sidh1999.debug-notebook

**Comment by zmbc** [comment]
@sidhanthapoddar99 I think [this limitation](https://marketplace.visualstudio.com/items/?itemName=sidh1999.debug-notebook):

> Text output only (no graphs or rich media)

Is effectively the problem we are trying to solve in this thread. It's a cool idea for other reasons though!

**Comment by sidhanthapoddar99** [comment]
@zmbc Fair point — and thanks for the push! That limitation was real when I first wrote it. After your comment I went back and reworked the rendering path: v2.0.1 (just shipped) now does inline rich output for 
- matplotlib.Figure, 
- pandas.DataFrame / Series, 
- plotly.Figure, PIL.Image, and 
- anything with _repr_html_.
The trailing expression of a cell auto-renders Jupyter-style, and plt.show() is monkey-patched to flush figures inline instead of opening a window.
  
  The interesting constraint was that everything has to round-trip through the DAP evaluate request — there's no kernel of our own. So rich payloads are emitted to a tempfile in the debuggee, the path comes back via the evaluate response, and the extension reads + unlinks. Took a few iterations to find a transport that doesn't get truncated by debugpy's internal repr-size caps and doesn't spam the debug console with base64.

Release notes + demo gif here: 
[https://github.com/sidhanthapoddar99/pyrept-notebook-vscode-debug extension/releases/tag/v2.0.1](https://github.com/sidhanthapoddar99/pyrept-notebook-vscode-debug-extension/releases/tag/v2.0.1)

Marketplace listing updates within the hour: 
https://marketplace.visualstudio.com/items?itemName=sidh1999.debug-notebook


<img width="2715" height="1760" alt="Image" src="https://github.com/user-attachments/assets/2711eb91-b1bc-48b2-9fd5-dc98b036884f" />

**Comment by RobinL** [comment]
@sidhanthapoddar99 I _really_ like that.  Up till now I've found it hard to articulate why this is such a useful thing, and I wasn't necessarily sure what the solution looked like, but I love the idea that the at debugging happens in a jupyter notebook, it seems like a perfect throwaway environment for exploring the current state of the program, with rich output and keeping everything copy pastable

**Comment by danielecandelaresi** [comment]
Hi all! Someone knows if this feature is under development?
